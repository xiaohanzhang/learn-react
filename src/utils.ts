export function debounce<T extends (...args: any) => any>(func: T, wait: number = 0) {
  let timeout: any;
  let result: any;

  return function debounced(this: any, ...args: any[]) {
    clearTimeout(timeout);
    timeout = setTimeout(() => {
      result = func.apply(this, args);
    }, wait);
    return result;
  };
}

const isEqualArrays = (src: any[], target: any[]) => {
  return src === target || (
    src.length === target.length && src.every((v, i) => v === target[i])
  );
};

const isEqualObjects = (src: Object, target: Object) => {
  if (src === target) {
    return true;
  }
  const srcKeys = Object.keys(src);
  const targetKeys = Object.keys(target);
  // @ts-ignore
  return srcKeys.length === targetKeys.length && srcKeys.every((k) => src[k] === target[k]);
}

export const isEqual = (src: any, target: any) => {
  if (src === target || (Number.isNaN(src) && Number.isNaN(target))) {
    return true;
  }
  const srcTag = Object.prototype.toString.call(src);
  const targetTag = Object.prototype.toString.call(target);
  return srcTag === targetTag && (
    (Array.isArray(src) && isEqualArrays(src, target)) ||
    (srcTag === '[object Object]' && isEqualObjects(src, target))
  )
}
