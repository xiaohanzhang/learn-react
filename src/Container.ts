import { isEqual, debounce } from './utils';

interface Props {
  key?: string,
}

export interface MyElement<P = {}> {
  type: string | FC,
  key: string | null,
  props: Props & P,
  children: MyChildren[],
}

type MyNode = string | number | boolean | null | undefined | MyElement;
export type MyChildren = MyNode | Array<MyNode | MyChildren>;
export type FC<P = {}> = (props: P & Props & {children?: any}, context?: any) => MyChildren;

const notNull = <T,>(value: T|null): value is T => {
  return value !== null;
}

interface HTMLProps {
  onClick?: (e: MouseEvent) => any
}
const updateHtmlElement = (element: HTMLElement, { onClick }: Props & HTMLProps): HTMLElement => {
  if (onClick) {
    element.onclick = onClick;
  }
  return element;
}

const createHtmlElement = (tag: string, props: Props & HTMLProps) => {
  const element = document.createElement(tag);
  return updateHtmlElement(element, props)
}

export const replaceChildren = (parent: Node, children: Node[]) => {
  if (children.length > 0) {
    children.forEach((child) => {
      parent.appendChild(child);
    });
    while (parent.firstChild !== null && parent.firstChild !== children[0]) {
      parent.removeChild(parent.firstChild);
    }
  }
  return parent;
}

const isText = (value: any): value is Text => {
  return value instanceof Node && value.nodeType === Node.TEXT_NODE;
}

class Container {
  static instance: Container | null = null;
  static getInstance() {
    return this.instance as Container;
  }

  element: HTMLElement | null = null;
  hookStates: any[] = [];
  hookStateIndex = 0;
  children: {[key: string]: Container|Text} = {};

  constructor(
    private prev: MyElement
  ) {}

  debounceRender = debounce((...args) => {
    return this.render(...args);
  })

  shouldRender(current: MyElement) {
    const { prev } = this;
    if (prev.type !== current.type) {
      this.destroy();
    }
    return !isEqual(prev.props, current.props);
  }

  destroy() {
    const { element } = this;
    if (element !== null) {
      element.remove();
      this.element = null;
    } else {
      Object.keys(this.children).forEach((key) => {
        this.destroyChild(key);
      });
    }
    this.children = {};
  }

  destroyChild(key: string) {
    if (key in this.children) {
      const child = this.children[key];
      if (isText(child)) {
        child.remove();
      } else {
        child.destroy();
      }
      delete this.children[key];
    }
  }

  getDirectNodes(): Node[] {
    if (this.element !== null) {
      return [this.element];
    }
    return Object.values(this.children).map((child) => {
      return isText(child) ? child : child.getDirectNodes();
    }).flat();
  }

  renderChildren(children: MyChildren[]): Node[] {
    const newChildren: {[key: string]: Container|Text} = {};
    const addChild = (key: string, child: Container|Text, destory: boolean) => {
      newChildren[key] = child;
      if (key in this.children) {
        delete this.children[key];
      }
      if (destory) {
        this.destroyChild(key);
      }
      return child;
    }
    const results = children.flat(Infinity)
      .map((child: MyNode, i) => {
        const key = `__learn_${i}`;
        // ignore boolean, null, undefined
        if (typeof child === 'string' || typeof child === 'number') {
          const node = document.createTextNode('' + child);
          addChild(key, node, true);
          return node;
        } else if (child && typeof child === 'object') {
          if (child.key === null) {
            child.key = key;
          }
          let container = this.children[child.key];
          let current: MyElement | null = child;
          if (!container || isText(container)) {
            container = new Container(child);
            current = null;
          }
          addChild(child.key, container, !current);
          return container.render(current);
        }
        return null;
      })
      .flat()
      .filter(notNull)
    ;
    Object.keys(this.children).forEach((key) => {
      this.destroyChild(key);
    });
    this.children = newChildren;
    return results;
  }

  render(current?: MyElement | null) {
    if (!current) {
      current = this.prev;
    } else if (!this.shouldRender(current)) {
      return this.getDirectNodes();
    }
    const { type, props, children } = current;
    if (typeof type === 'string') {
      if (this.element === null) {
        this.element = createHtmlElement(type, props);
      } else {
        this.element = updateHtmlElement(this.element, props);
      }
      return [replaceChildren(this.element, this.renderChildren(children))];
    } else {
      Container.instance = this;
      this.hookStateIndex = 0;
      const newChildren = [type({...props, children})];
      return this.renderChildren(newChildren);
    }
  }

  nextHookState<T>(initialState: T): [T, boolean] {
    const isNew = this.hookStateIndex === this.hookStates.length;
    if (isNew) {
      this.hookStates.push(initialState);
    }
    const state = this.hookStates[this.hookStateIndex];
    this.hookStateIndex += 1;
    return [state, isNew];
  }

  useRef<T>(initial: T): {current: T} {
    return this.nextHookState({current: initial})[0];
  }

  useState<T>(initial: T): [T, (value: T) => T] {
    const [state] = this.nextHookState<{
      value: T, setValue: null | ((value: T) => T)
    }>({ value: initial, setValue: null });

    if (state.setValue === null) {
      state.setValue = (value: T) => {
        state.value = value;
        this.debounceRender();
        return value;
      };
    }

    return [state.value, state.setValue];
  }
}

export default Container;
