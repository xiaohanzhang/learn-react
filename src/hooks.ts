import Container from './Container';

export const useRef = <T,>(initial: T): {current: T} => {
  return Container.getInstance().useRef(initial);
};

export const useState = <T,>(initial: T) => {
  return Container.getInstance().useState(initial);
};

export const useReducer = () => {
  throw new Error('not implemented');
};

export const useContext = () => {
  throw new Error('not implemented');
}
