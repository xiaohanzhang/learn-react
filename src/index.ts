import Container, { FC, MyChildren, MyElement, replaceChildren } from './Container';
export { useState, useRef } from './hooks';

export const createElement = <P extends {}>(
  type: FC | string, 
  props: {key?: string | number} & P | null = null, 
  ...children: MyChildren[]
): MyElement => {
  const { key, ...rest } = props || {};
  return {
    type,
    key: key !== undefined ? '' + key : null,
    props: rest,
    children,
  };
}

export const render = (current: MyElement, element: Element | null) => {
  if (element === null) {
    return;
  }
  // @ts-ignore 
  let container: Container = element.__learnRoot;
  if (!container) {
    // @ts-ignore 
    container = element.__learnRoot = new Container(current);
    while (element.lastChild) {
      element.removeChild(element.lastChild);
    }
    replaceChildren(element, container.render());
  } else {
    replaceChildren(element, container.render(current));
  }
  return element;
}
