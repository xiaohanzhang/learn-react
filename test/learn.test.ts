import { render, createElement as ce } from '../src';

describe('leart-react', () => {
  it('renders single div', () => {
    const div = document.createElement('div');
    expect(render(ce('div', null, 'hello world'), div)).toMatchSnapshot();
  });
});
