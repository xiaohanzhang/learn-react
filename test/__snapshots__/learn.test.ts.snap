// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leart-react renders single div 1`] = `
<div>
  <div>
    hello world
  </div>
</div>
`;
